GH-33346: [Python] DataFrame Interchange Protocol for pyarrow Table

AlenkaF opened the pull request:
This PR implements the Dataframe Interchange Protocol for `pyarrow.Table`.
See: https://data-apis.org/dataframe-protocol/latest/index.html

AlenkaF commented on python/pyarrow/interchange/from_dataframe.py:
```diff
@@ -0,0 +1,529 @@
+from pyarrow.interchange.column import Dtype
+
+
+# A typing protocol could be added later to let Mypy validate code using
+# `from_dataframe` better.
+DataFrameObject = Any
+ColumnObject = Any
+BufferObject = Any
```
When I removed copied version of the [dataframe protocol](https://github.com/data-apis/dataframe-api/blob/aa6fe7d7bc4fd6fd24b8dd6b4dfb8c58cac2d8b9/protocol/dataframe_protocol.py#L275) I also lost the definition of base classes for `Buffer`, `Column` and `DataFrame`. I substituted the base classes with `Any` type hint. 

For reference, pandas uses dataframe protocol base class with copied protocol spec, Vaex uses `Any` type hints as we do here.

AlenkaF commented on python/pyarrow/interchange/from_dataframe.py:
```diff
@@ -0,0 +1,529 @@
+        return unit, tz
+
+    # TODO
+    # date 'td{Days/Ms}'
```
Currently I implemented the protocol for `timestamp` dtype (for consuming and producing) and not for `date` or `time.` I do plan to test and add support for those data types also (can do in a separate PR).

jorisvandenbossche replied:
I think it is fine to only support timestamp. I assume the protocol doesn't actually support `date` and `time`, although `date` could be seen as datetime with day resolution (which you can represent with numpy's datetime64 dtype), but I don't know if that is officially allowed by the protocol.

AlenkaF replied:
Thanks for the feedback. Will leave it as is for now.

AlenkaF commented on python/pyarrow/tests/interchange/test_extra.py:
```diff
@@ -0,0 +1,221 @@
+        from_dataframe as pandas_from_dataframe
+    )
+
+    data = np.array([b'x'*1024]*(3*1024**2), dtype='object')  # 3GB bytes data
```
I think the [failing tests](https://github.com/apache/arrow/actions/runs/3682323841/jobs/6229859945) have something to do with the LargeStringArray. Will test it on the CI later with using smaller array.

AlenkaF commented:
One more thing I will try to look at asap is the copies being made in the implementation and will add warnings in case `allow_copy=False`.

jorisvandenbossche commented on python/pyarrow/interchange/__init__.py:
```diff
@@ -0,0 +1,16 @@
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
```
I think we can expose the `from_dataframe` function here in the init (so you don't have to import it from the sub-submodule)

jorisvandenbossche reviewed (commented):
(partial review)

AlenkaF commented on python/pyarrow/interchange/column.py:
```diff
@@ -0,0 +1,507 @@
+if sys.version_info >= (3, 8):
+    from typing import TypedDict
+else:
+    from typing_extensions import TypedDict
```
I think it is part of the Python standard library: https://docs.python.org/3/library/typing.html.
Looking at it `typing` should be supported for versions 3.5 and up. Will change the check.

AlenkaF commented on python/pyarrow/interchange/column.py:
```diff
@@ -0,0 +1,507 @@
+
+class DtypeKind(enum.IntEnum):
+    """
+    Integer enum for data types.
+    Attributes
```
Yes, the whitespace lines were lost and I guess I haven't found all of them =) Will go through them again!

AlenkaF commented on python/pyarrow/interchange/column.py:
```diff
@@ -0,0 +1,507 @@
+    pa.float64(): (DtypeKind.FLOAT, "g"),
+    pa.bool_(): (DtypeKind.BOOL, "b"),
+    pa.string(): (DtypeKind.STRING, "u"),  # utf-8
+    pa.large_string(): (DtypeKind.STRING, "U"),
```
Yes

https://github.com/apache/arrow/blob/71ca596d233195105466579d138cf61df81a6a89/python/pyarrow/tests/interchange/test_extra.py#L172-L187

but I did have to change the `format_string`:

https://github.com/apache/arrow/blob/71ca596d233195105466579d138cf61df81a6a89/python/pyarrow/interchange/column.py#L268-L272

What we talked about and I should note and add to the tests is that when consuming a string array from pandas it will always result in a LargeStringArray due to pandas offset being int64.

AlenkaF commented on python/pyarrow/interchange/column.py:
```diff
@@ -0,0 +1,507 @@
+    is_dictionary: bool
+    # Python-level only (e.g. ``{int: str}``).
+    # None if not a dictionary-style categorical.
+    # categories: Optional[Column]
```
Hm, I do not remember why I commented it out. Will revert and see.

AlenkaF commented on python/pyarrow/interchange/column.py:
```diff
@@ -0,0 +1,507 @@
+                )
+                i += 1
+
+        elif isinstance(self._col, pa.ChunkedArray):
```
> Above you combined the chunks when passing a ChunkedArray to the constructor, so that means self._col can currently never be a ChunkedArray?

Oh, I missed this part here when moving the combining of chunks to the constructor in https://github.com/apache/arrow/pull/14804/commits/71ca596d233195105466579d138cf61df81a6a89
Will correct.

> But maybe we should try to support that and not always upfront call combine_chunks?

Just to summarize what we talked about (will add it to the code as a comment): when the dataframe is being consumed with `from_dataframe` method it iterates through the chunks of the dataframe: https://github.com/apache/arrow/blob/71ca596d233195105466579d138cf61df81a6a89/python/pyarrow/interchange/from_dataframe.py#L104 https://github.com/apache/arrow/blob/71ca596d233195105466579d138cf61df81a6a89/python/pyarrow/interchange/dataframe.py#L161-L190 
and so is always converting part of the dataframe that is not chunked. For that reason there is no need to support chunking on the column level as it is only possible to get a chunked array if calling Column class directly, not through the DataFrame class.

AlenkaF commented on python/pyarrow/interchange/dataframe.py:
```diff
@@ -0,0 +1,190 @@
+        # producer to overwrite null values in the data with ``NaN`` (or
+        # ``NaT``). This currently has no effect; once support for nullable
+        # extension dtypes is added, this value should be propagated to
+        # columns.
```
It is coming from the protocol spec: https://github.com/data-apis/dataframe-api/blob/aa6fe7d7bc4fd6fd24b8dd6b4dfb8c58cac2d8b9/protocol/dataframe_protocol.py#L400-L403

Agree, it would make sense to add a warning that the keyword doesn't have an effect. Will do.

jorisvandenbossche replied:
Yes, but the "once support for nullable extension dtypes is added, this value should be propagated to columns." part of the comment is pandas specific, and can be removed I think

AlenkaF commented on python/pyarrow/interchange/dataframe.py:
```diff
@@ -0,0 +1,190 @@
+        """
+        Return the number of chunks the DataFrame consists of.
+        """
+        return self._df.column(0).num_chunks
```
> With the current implementation, this is always 1 ?

No, that shouldn't be the case. With the dataframe interchange protocol class, Column always gets combined into one chunk. But here we are calling PyArrow array which is not a part of the dataframe interchange protocol implementation so we will always get the correct result of the chunks.

```python
>>> table = pa.table([pa.chunked_array([[2, 2, 4], [4, 5, 100]])], names=["Chunked"])
>>> table
pyarrow.Table
Chunked: int64
----
Chunked: [[2,2,4],[4,5,100]]
>>> table_interchange = table.__dataframe__()


>>> table.column(0).num_chunks
2
>>> table_interchange.get_column(0).num_chunks()
1
```

I will test that when adding the `PyArrow <-> PyArrow` roundtrips.

> (it's also not super clear what the "number of chunks" for a DataFrame actually is, if each column can have a different number of chunks ..)

Oh, haven't thought of that. Looking at the design concept I think the protocol implies dataframe and its columns are chunked in the same way, see
https://data-apis.org/dataframe-protocol/latest/design_requirements.html?highlight=validity#conceptual-model-of-a-dataframe

And I took the first column of the dataframe as the number of chunks for all other columns.

AlenkaF commented:
@jorisvandenbossche I think I addressed all the comments and topics we have talked about. Some information about the recent changes:

- Large string dtype is working correctly. Pandas doesn't support this type of data so I changed the adaptation and added a test to check for error due to not supported dtype. PyArrow roundtrip works correctly - I had to construct an array with `pyarrow.LargeStringArray.from_buffers` and not `pyarrow.Array.from_buffers` in case of large string dtype.
- String dtype is removed from pandas roundtrip tests as pandas defines `.size()` as a method in [column.py](https://github.com/pandas-dev/pandas/blob/91111fd99898d9dcaa6bf6bedb662db4108da6e6/pandas/core/interchange/column.py#L84-L88) but calls it as a property in [from_dataframe.py](https://github.com/pandas-dev/pandas/blob/5c66e65d7b9fef47ccb585ce2fd0b3ea18dc82ea/pandas/core/interchange/from_dataframe.py#L247) and so the roundtrip with pandas errors for string dtypes.
- I have added better test coverage with parametrizing some of the tests and also using strategies for one of the test in `test_interchange_spec.py`.
- Now an error is raised if `nan_as_null=True`, I also added a test for it.
- I have exposed `from_dataframe` in `interchange/__init__.py`, hope I have done it correctly.

So I think this PR is ready for another round of review 🙏

jorisvandenbossche commented on python/pyarrow/interchange/from_dataframe.py:
```diff
@@ -0,0 +1,590 @@
+                "PyArrow Array will always be created out of a NumPy ndarray "
+                "and a copy is required which is forbidden by allow_copy=False"
+            )
+        return pa.array(array.to_pylist(), mask=bytemask), buffers
```
We should make this easier to create a new array from an existing array + new validity bitmap or mask, without the need to go through `to_pylist`. https://issues.apache.org/jira/browse/ARROW-7071 might be a related feature request.

Although in this case, doing `array.to_numpy()` might be OK as well? (conversion to numpy should be cheaper than to a python list)

jorisvandenbossche commented on python/pyarrow/interchange/from_dataframe.py:
```diff
@@ -0,0 +1,590 @@
+    return validity_pa_buff, bytemask
+
+
+def buffer_to_ndarray(
```
Why do we need this helper for a bytemask? I would assume we can take the same approach as for int8 arrays? (which would go through pa.foreign_buffer and pa.Array.from_buffers)

jorisvandenbossche commented on python/pyarrow/tests/interchange/test_extra.py:
```diff
@@ -0,0 +1,371 @@
+        }
+    )
+
+    from pandas.core.interchange.from_dataframe import (
+        from_dataframe as pandas_from_dataframe
```
You can import this from `pandas.api.interchange` (that is considered the public exposed location)

jorisvandenbossche commented on python/pyarrow/tests/interchange/test_extra.py:
```diff
@@ -0,0 +1,371 @@
+    "float, np_float", [
+        (pa.float16(), np.float16),
+        # (pa.float32(), np.float32),   #errors due to inequality in nan
+        # (pa.float64(), np.float64)    #errors due to inequality in nan
```
What's the error exactly?

jorisvandenbossche commented on python/pyarrow/tests/interchange/test_extra.py:
```diff
@@ -0,0 +1,371 @@
+)
+@pytest.mark.parametrize("unit", ['s', 'ms', 'us', 'ns'])
+@pytest.mark.parametrize("tz", ['America/New_York', '+07:30', '-04:30'])
+def test_pyarrow_roundtrip(uint, int, float, np_float, unit, tz):
```
In this one we can maybe parametrize it for being sliced as well (eg `table = table.slice(1, 1)` before doing the roundtrip)

jorisvandenbossche commented on python/pyarrow/tests/interchange/test_extra.py:
```diff
@@ -0,0 +1,371 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
Maybe call this file something like `test_conversion.py`? (more descriptive than `test_extra.py`, and I think this file is mostly checking the actual conversion, and the other file is more testing the API of the spec object?)

jorisvandenbossche reviewed (commented):
Great work! Continuing my review (now mostly looked at `from_dataframe`)

jorisvandenbossche commented on python/pyarrow/interchange/from_dataframe.py:
```diff
@@ -0,0 +1,590 @@
+    if offset_buff:
+        # If an offset buffer exists, construct an offset pyarrow Buffer
+        # and add it to the construction of an array
+        offset_pa_buffer = pa.py_buffer(offset_buff._x)
```
Avoid accessing the private `_x` ?

jorisvandenbossche commented on python/pyarrow/tests/interchange/test_extra.py:
```diff
@@ -0,0 +1,371 @@
+    if Version(pd.__version__) < Version("1.5.0"):
+        pytest.skip("__dataframe__ added to pandas in 1.5.0")
+
+    arr = [1, 2, 3]
```
To ensure to have coverage of missing values, maybe add a separate test that just does pandas->pyarrow (so creating the pandas DataFrame, without doing the pyarrow->pandas part of the roundtrip, which currently fails)

AlenkaF commented on python/pyarrow/interchange/from_dataframe.py:
```diff
@@ -0,0 +1,590 @@
+        else:
+            raise NotImplementedError(f"Data type {dtype} not handled yet")
+
+        buffers.append(buf)
```
Ha, I have missed that. I see that Vaex also saves the list of all buffers to a private attribute `_buffers` of the resulting dataframe. Should we do the same, add all the buffers to a private attribute of the created table? Or passing separate buffers to the `foreign_buffer` functions should suffice?

It might be good to have the buffers saved in the private attribute of the table as they can be easily inspected if needed?

jorisvandenbossche replied:
I think it is better (and easier) to use the `base` keyword in foreign_buffer for that. We would otherwise have to add a specific attribute for this (Table being a cython class doesn't allow dynamically adding attributes), and I think it is also more robust to tie it to the buffer itself (the table might go out of scope if the user for example slices it).

AlenkaF replied:
Makes sense, thank you for the explanation 🙏

AlenkaF commented:
@jorisvandenbossche I have gone through all of the suggestions, I think this PR is ready for another round of review. Thank you!

jorisvandenbossche commented on python/pyarrow/interchange/from_dataframe.py:
```diff
@@ -0,0 +1,540 @@
+def protocol_df_chunk_to_pyarrow(
+    df: DataFrameObject,
+    allow_copy: bool = True
+) -> pa.Table:
```
```suggestion
) -> pa.RecordBatch:
```

jorisvandenbossche commented on python/pyarrow/interchange/from_dataframe.py:
```diff
@@ -0,0 +1,540 @@
+    """
+    if not allow_copy:
+        raise RuntimeError(
+            "Boolean column will be casted from uint8 and a copy "
+            "is required which is forbidden by allow_copy=False"
```
This error message is about boolean column, but should be about dictionary/categorical

jorisvandenbossche commented on python/pyarrow/interchange/from_dataframe.py:
```diff
@@ -0,0 +1,540 @@
+            DtypeKind.STRING,
+            DtypeKind.DATETIME,
+        ):
+            columns[name] = column_to_array(col)
```
Should `allow_copy` be passed here as well? Also primitive arrays can result in a copy, I think, in case that the null type doesn't match pyarrow?

AlenkaF replied:
Hm, I am not sure I agree (but am certain you are never wrong 😄 ). If null type doesn't match pyarrow then the mask is created from the data but then it gets added as a validity buffer in `pa.Array.from_buffers` but data buffer stays the same to the one of the primitive array from input.

But will try an example locally to see the pointers of the buffers.

jorisvandenbossche replied:
Ah, that's true. It's a bit ambiguous, the data itself isn't copied, but the conversion is also not fully what I would consider "zero copy" as you still need to create new data. Anyway, that seems like a corner case that is fine to ignore right now.

jorisvandenbossche commented on python/pyarrow/interchange/from_dataframe.py:
```diff
@@ -0,0 +1,540 @@
+    # Check for float NaN values
+    if null_kind == ColumnNullType.USE_NAN:
+        if kind == DtypeKind.FLOAT and bit_width == 16:
+            raise NotImplementedError(
+                f"{data_type} with {null_kind} is not yet supported.")
```
What was the reason again that this doesn't work? (maybe add a comment about it)

jorisvandenbossche replied:
Ah, I see in the tests that there is a comment saying the `is_nan` compute kernel isn't yet implemented for float16

AlenkaF replied:
Yes, will add the same comment to the code also so that it is clearer.

jorisvandenbossche commented on python/pyarrow/interchange/from_dataframe.py:
```diff
@@ -0,0 +1,540 @@
+            sentinel_dtype = pa.int64()
+        else:
+            sentinel_dtype = data_dtype
+        int_arr = pa.Array.from_buffers(sentinel_dtype,
+                                        length,
+                                        [None, data_pa_buffer],
+                                        offset=offset)
+        sentinel_arr = pc.equal(int_arr, sentinel_val)
```
```suggestion
        pyarrow_data = pa.Array.from_buffers(sentinel_dtype,
                                        length,
                                        [None, data_pa_buffer],
                                        offset=offset)
        sentinel_arr = pc.equal(pyarrow_data, sentinel_val)
```

(it's not necessarily an int array?)

jorisvandenbossche commented on python/pyarrow/interchange/from_dataframe.py:
```diff
@@ -0,0 +1,540 @@
+                                             offset_buff.bufsize,
+                                             base=offset_buff)
+
+        if data_type[2] == 'U':
+            if not null_count:
+                null_count = -1
+            array = pa.LargeStringArray.from_buffers(
+                length,
+                offset_pa_buffer,
+                data_pa_buffer,
+                validity_pa_buff,
+                null_count=null_count,
+                offset=offset,
+            )
+        else:
+            if offset_bit_width == 64:
+                string_type = pa.large_string()
+            else:
+                string_type = pa.string()
+            array = pa.Array.from_buffers(
+                string_type,
+                length,
+                [validity_pa_buff, offset_pa_buffer, data_pa_buffer],
+                offset=offset,
+            )
```
Can we combine the if/else here? (and do a single pa.Array.from_buffers) 

I think you could do something like `if data_type[2] == "U": string_type = pa.large_string()`

jorisvandenbossche commented on python/pyarrow/interchange/from_dataframe.py:
```diff
@@ -0,0 +1,540 @@
+                                             base=offset_buff)
+
+        if data_type[2] == 'U':
+            if not null_count:
+                null_count = -1
```
This is not needed in the `else: ` branch? (or, what is the reason it is needed here?)

And should the check be `if null_count is None:`? (with `if not null_count`, also the case of null_count=0 would set it to -1)

AlenkaF replied:
Thinking about it now this is not needed at all. Will remove.

jorisvandenbossche reviewed (commented):
> * String dtype is removed from pandas roundtrip tests as pandas defines `.size()` as a method in [column.py](https://github.com/pandas-dev/pandas/blob/91111fd99898d9dcaa6bf6bedb662db4108da6e6/pandas/core/interchange/column.py#L84-L88) but calls it as a property in [from_dataframe.py](https://github.com/pandas-dev/pandas/blob/5c66e65d7b9fef47ccb585ce2fd0b3ea18dc82ea/pandas/core/interchange/from_dataframe.py#L247) and so the roundtrip with pandas errors for string dtypes.

It should be possible to add this back now (since it is fixed in pandas main), but skip the string dtype depending on the pandas version (only the 2.0.0.dev version runs the test)

jorisvandenbossche commented:
@github-actions crossbow submit test-conda-python-3.9-pandas-upstream_devel

jorisvandenbossche commented:
@github-actions crossbow submit test-conda-python-3.8-pandas-nightly

AlenkaF commented:
@jorisvandenbossche the code review suggestions are all addressed. If I am not mistaken you still want to review the tests?

jorisvandenbossche commented:
@github-actions crossbow submit test-conda-python-3.8-pandas-nightly

jorisvandenbossche commented:
@github-actions crossbow submit test-conda-python-3.8-pandas-nightly

AlenkaF commented:
@github-actions crossbow submit test-conda-python-3.8-pandas-nightly

jorisvandenbossche commented on python/pyarrow/tests/interchange/test_conversion.py:
```diff
@@ -192,6 +187,34 @@ def test_pandas_roundtrip(uint, int, float, np_float):
+    )
+    pandas_df = pandas_from_dataframe(table)
+    result = pi.from_dataframe(pandas_df)
+
```
Is there a `assert table.equals(result)` missing here (like there is in the test above)?

AlenkaF replied:
Due to pandas defining `int64` offset for what is in our case normal string, not large, the dtype that is at the end of the roundtrip becomes `large_string`. Due to that, the assertion is done with pylist for the vakues and separate for dtype (first inormal string, then large string).

AlenkaF commented:
@jorisvandenbossche do you have any blocking issues I can correct today to make this PR merged before the release freeze? Hope the answer to the question about assertions on `test_roundtrip_pandas_string ` was clear enough.

rok commented:
Great to see this merged!
